File: analytics/define_encoder.py
import numpy as np
from activations import *

def foward_pass(X, W, B):
    Y = np.dot(X, W) + B
    return Y

def forward_prop(Layers, labels):
    x = Layers[0]['x']

    for i, layer in enumerate(Layers):
        w = layer['w']
        b = layer['b']
        actv = layer['activation']
        Y = foward_pass(x, w, b)

        if actv is not 'relu':
            x = mse(Y, labels)
        else:
            x = activation(Y, actv)
    return x

def add(layers : list, num_nodes : int, activation : str):
    w_prev_size = layers[-1]['w'].shape[1]
    w_current = np.random.randn(w_prev_size, num_nodes)
    layers.append({
        'b': 1,
        'w': w_current,
        'activation': activation
    })
    return layers

if __name__ == '__main__':
    layers = [
        {
            'x': np.random.randn(1, 3),
            'b': 1,
            'w': np.random.randn(3, 2),
            'activation': 'relu',
        },
    ]

    labels = np.random.randn(1, 3)
    layers = add(layers, 2, 'relu')
    layers = add(layers, 1, 'relu')
    layers = add(layers, 2, 'relu')
    layers = add(layers, 3, 'mse')

    output = forward_prop(layers, labels)
    print(labels)

    print(output)

File: analytics/define_encoder.cpp
#include <iostream>
#include <map>
#include <vector>
#include <Eigen/Dense>

using namespace std;
using namespace Eigen;

auto forward_pass(MatrixXf X, MatrixXf W) {
    auto Y = X * W;
    return Y;
}

auto forward_prop(vector<map<string, MatrixXf>> Layers, vector<string> activations, MatrixXf labels){
    for (auto layer : Layers) {
        auto w = layer["w"];
        cout<<w;
    }
    return 1;
}

int main(){
    MatrixXf x(1, 3);
    x << 1, 3, 4;

    MatrixXf w(3, 2);
    w << 2.3, 3.4, 5.4,
         4.5, 6.7, 2.3;
         

    vector<map<string, MatrixXf>> Layers;
    map<string, MatrixXf> m;
    m["w"] = w;
    Layers.push_back(m);

    vector<string> activations;
    activations.push_back("relu");
    
    MatrixXf labels(1, 1); 
    labels << 1;

    cout<<forward_prop(Layers, activations, labels);
}